### PR TITLE
feat : Add User Authentication

### LINKED ISSUE #363: User Authentication in the app
We would want to have login functionality but the app features shouldn't depend on that. So, if a user logs in, he gets a personalised experience with his custom settings but if he doesn't login he gets to see all the basic information and features like talks, bookmark sessions etc.
- [ ] Google+ sign in
- [ ] Facebook Oauth
- [ ] Twitter Oauth
- [ ] Server only login ( This is more inclined for the f droid version where we can't be using the above three login mechanisms )
- [ ] No login at all ( Skip login )

### PR DESCRIPTION
**Pull request by Shailesh351**
Fixes #363 

Changes:

- [x] Add login activity
- [x] Add Login model and setup login API service 
- [x] Add Sign up(new user) activity
- [x] Add email and password field validation
- [x] Add User model and setup signup(new user) API service
- [x] Save token in shared preference and add JWT Utils
- [x] Define final behaviour(flow) and UI
- [x] Add logout option 
- [x] Reuse code to minimize code base. 
- [x] Update Auth to JSONAPI

### CODE REVIEW

**Comment by Shailesh351** [comment]
Sign Up Demo
signup

Login Demo
signin

@iamareebjamal @arp95 @mahikaw @rohanagarwal94 @shreyasnbhat Please test and review.
@mananwason @the-dagger Please review.

Almost all the work for user authentication is done. Log out is remaining. I'll add it soon.

**Comment by Shailesh351** [comment]
@niranjan94  Should we add option for isAuthEnabled in config.json ? According to it app will show auth feature.

**Comment by niranjan94** [comment]
> Should we add option for isAuthEnabled in config.json ? According to it app will show auth feature.

@Shailesh351 sounds good. Add also, this needs to be given as an option in the generator. (But that can come in a separate PR)

**Comment by Shailesh351** [comment]
@iamareebjamal @arp95 @mahikaw @rohanagarwal94 @shreyasnbhat Please test and review :smile:

File: android/app/src/main/assets/config.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
   "Email": "dev@fossasia.org",
   "App_Name": "Open Event",
-  "Api_Link": "https://eventyay.com/api/v1/events/6/"
+  "Api_Link": "http://open-event-dev.herokuapp.com/api/v1/events/6/"
```
Comment (iamareebjamal) [inline android/app/src/main/assets/config.json]: No such event exists and use https

File: android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+                APIClient.getOpenEventAPI().login(new Login(email, password))
+                        .subscribeOn(Schedulers.io())
+                        .observeOn(AndroidSchedulers.mainThread())
+                        .subscribe(loginResponseObserver);
+            }
+        });
+    }
+
+    private void setupLoginResponseObserver() {
+        loginResponseObserver = new Observer<LoginResponse>() {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java]: Why creating a new Observer? Subscribe and pass the appropriate components

File: android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        } else {
+            progressBar.setVisibility(View.GONE);
+        }
+    }
+
+    @Override
+    protected void onStop() {
+        super.onStop();
+
+    }
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java]: This isn't needed if no action is triggered here

File: android/app/src/main/java/org/fossasia/openevent/activities/SignUpActivity.java [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+                APIClient.getOpenEventAPI().signUp(new SignUp(email, password))
+                        .subscribeOn(Schedulers.io())
+                        .observeOn(AndroidSchedulers.mainThread())
+                        .subscribe(signUpResponseObserver);
+            }
+        });
+    }
+
+    private void setupSignUpResponseObserver() {
+        signUpResponseObserver = new Observer<User>() {
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/SignUpActivity.java]: Same thing about observer

File: android/app/src/main/java/org/fossasia/openevent/activities/UserProfileActivity.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            logoutDialog.show();
+        });
+    }
+
+    private void logout() {
+        SharedPreferencesUtil.remove(ConstantStrings.TOKEN);
+        SharedPreferencesUtil.remove(ConstantStrings.USER_EMAIL);
+        goToMainActivity();
+        Toast.makeText(UserProfileActivity.this, getString(R.string.logged_out_successfully), Toast.LENGTH_SHORT).show();
+        Timber.d("Removes token & email and logged out successfully");
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/activities/UserProfileActivity.java]: Removed

**Review by iamareebjamal — CHANGES_REQUESTED** [review]

File: android/app/src/main/assets/config.json (hunk 1 above)
Comment (Shailesh351) [reply android/app/src/main/assets/config.json]: I changed it to test it on development deployment. Actually i forgot to change it back. Will change it :+1:

File: android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java (hunk 2 above)
Comment (Shailesh351) [reply android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java]: You mean using doOnSubscribe(), doOnNext(), etc.. ?

Comment (iamareebjamal) [reply android/app/src/main/java/org/fossasia/openevent/activities/LoginActivity.java]: No. subscribe(data -> { // onNext }, throwable -> { // error }, () -> { // completed })

**Comment by Shailesh351** [comment]
@iamareebjamal Made changes please review now

File: android/app/src/main/java/org/fossasia/openevent/api/network/OpenEventAPI.java [hunk 6]
```diff
@@ -20,6 +27,12 @@
  */
 public interface OpenEventAPI {
 
+    @POST("../../users")
+    Observable<User> signUp(@Body SignUp signUp);
+
+    @POST("../../login")
+    Observable<LoginResponse> login(@Body Login login);
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/api/network/OpenEventAPI.java]: This is not the correct endpoint. It should be `/auth/session` I guess

File: android/app/src/main/java/org/fossasia/openevent/data/auth/LoginResponse.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonNaming;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonNaming(PropertyNamingStrategy.SnakeCaseStrategy.class)
```
Comment (iamareebjamal) [inline android/app/src/main/java/org/fossasia/openevent/data/auth/LoginResponse.java]: I think you are creating auth for older API. Please develop for JSONAPI

**Review by iamareebjamal — CHANGES_REQUESTED** [review]
I think it'll be better if you continue this after JSONAPI PR is completed

**Comment by Shailesh351** [comment]
@iamareebjamal @mahikaw @rohanagarwal94 @shreyasnbhat @arp95 Please review :smile:

**Comment by Shailesh351** [comment]
@iamareebjamal @rohanagarwal94 @shreyasnbhat @arp95 Guys please review :smile:

**Review by iamareebjamal — APPROVED** [review]

**Review by rohanagarwal94 — APPROVED** [review]

**Pull request merged**